### PR TITLE
add support for mlflow

### PR DESCRIPTION
**Pull request by mirkobronzi**
* add GPU support
* add mlflow support

### CODE REVIEW

File: dl_training/model/model_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        :param saved_model_path: if not None, will save models here
+        :param eval_smaller_is_better: if True, will try to minimize the eval metric.
+                                       if False, will try to maximize
+        :param patience: patience for early stopping
+        :param max_epoch: upper bound on the number of epoch
+        :return: the best eval value
+        """
+
+        not_improving_for = 0
+        best_eval_value = 9999999 if eval_smaller_is_better else -9999999  # big number
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: Instead of arbitrary high values, you can use positive or negative infinity like [-]numpy.inf (https://www.numpy.org/devdocs/reference/constants.html) or [-]float("inf")

File: dl_training/model/model_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                if saved_model_path is None:
+                    message = 'new best eval value - not saving models'
+                else:
+                    message = 'new best eval value - saving models to {}'.format(saved_model_path)
+                    torch.save(self.model.state_dict(), saved_model_path)
+            else:
+                not_improving_for += 1
+                message = "not improving since {} epoch".format(not_improving_for)
+
+            logger.info('epoch {:4} - loss: {:3.3f} - (dev) wer: {:3.3f} => {}'.format(
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: It feels weird to have the eval function softcoded but it's name "wer" being hardcoded in the message above.

File: dl_training/model/model_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                i, loss, eval_value, message))
+
+            log_metric("wer", best_eval_value)
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: Same here about the "wer" name being hardcoded.

File: dl_training/model/model_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        :param inputs: list of sentences. every sentence is a list of token inddices.
+        :param targets: list of sentences. every sentence is a list of token inddices.
+        :return: the average loss.
+        """
+
+        self.model.train()
+
+        losses = []
+        for batch_index in range(int(len(inputs) / self.batch_size)):
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: Computing the number of minibatches this way may omit some training examples if the size of the training set is not a multiple of self.batch_size.
Ex : len(inputs) == 10 and batch_size = 3
int(len(inputs) / batch_size) = int(10/3) = 3 minibatches of size 3 (the 10th training example is omitted).

File: dl_training/model/model_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            input_batch = inputs[
+                          batch_index * self.batch_size:(batch_index + 1) * self.batch_size]
+            target_batch = targets[
+                           batch_index * self.batch_size:(batch_index + 1) * self.batch_size]
+            loss = self.train_one_batch(input_batch, target_batch)
+            losses.append(loss)
+        return sum(losses) / len(losses)
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: If you adapt the code above not to ignore leftover training examples, this will need to be adapted to consider the fact the minibatches are going to start having different sizes.

File: dl_training/model/model_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        :return: the predictions.
+        """
+
+        self.model.eval()
+
+        if batch_size is None:
+            batch_size = self.batch_size
+
+        result = []
+        for batch_index in range(int(len(inputs) / batch_size)):
```
Comment (carriepl) [inline dl_training/model/model_wrapper.py]: Same issue with leftover examples being ignored. Passing 10 examples with batch_size==3 will return 9 predictions.

Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: It is preferred that batch_size is outside of the scope of the __init__ to allow dynamic batching of different sizes. Instead of using batch_size there you can infer the sizes of the inputs/targets in other methods.

File: dl_training/model/model_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    def predict_one_batch(self, input_batch):
+        """
+        run predition on the provided batch.
+
+        :param input_batch: list of sentences. every sentence is a list of token indices.
+        :return: the output predictions.
+        """
+
+        inputs = self.get_tensor_and_batch(input_batch)
+        result = self.model.predict(inputs).detach().to('cpu').numpy()
```
Comment (FrancisGregoire) [inline dl_training/model/model_wrapper.py]: Add a `with torch.no_grad()` score before calling `self.model.predict()`.

File: dl_training/model/model_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        return result.tolist()
+
+    def get_tensor_and_batch(self, xs):
+        longest = max([len(x) for x in xs])
+        padded = []
+        for x in xs:
```
Comment (FrancisGregoire) [inline dl_training/model/model_wrapper.py]: Should padding be done outside of this wrapper (e.g. in the dataloader using a collate function)?

Comment (carriepl) [reply dl_training/model/model_wrapper.py]: One advantage of doing the padding once you know what the minibatch will contain (instead of when you first load the data) is that you will, on average, need less padding so training will be faster.

File: dl_training/model/model_wrapper.py (hunk 4 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: yes, I wanted to take care of this later on - and now it's a good time..

File: dl_training/model/model_wrapper.py (hunk 6 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: do you mean we should remove `batch_size` from `__init__`?
that would force the user to specify the batch size every time we call `train` or `predict`

Assuming this is your point (may be wrong), I would prefer the way it is now to avoid repeating the batch size every time we call train..
@FrancisGregoire : what do you think?

File: dl_training/model/model_wrapper.py (hunk 8 above)
Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: Agreed. This is what I meant by dataloader/batcher.

File: dl_training/model/model_wrapper.py (hunk 7 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: do we need it also if we set `model.eval()` ?
(which is set by the method called before)

File: dl_training/model/model_wrapper.py (hunk 6 above)
Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: When training/predicting your `batch_size` is not static. Like @carriepl experienced, when your last batch is the leftover of your dataset your `batch_size` will be different (smaller). `inputs.size(0)` or other tricks should do the work.

File: dl_training/model/model_wrapper.py (hunk 1 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: indeed, fixed - inf is better :)

File: dl_training/model/model_wrapper.py (hunk 2 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: right - refactoring problems.. fixed now

File: dl_training/model/model_wrapper.py (hunk 3 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: also fixed

File: dl_training/model/model_wrapper.py (hunk 6 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: yes indeed, I will deal with the problems @carriepl mentioned before.
but before doing it, the point is if this code should have the responsibility to create the batches.

If yes, then I will fix.
If not, then I will just remove all the batching/padding code from here.

But we need to decide that before..

File: dl_training/model/model_wrapper.py (hunk 7 above)
Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: Yes and eval will be similar, but it needs to accumulate a loss.

File: dl_training/model/model_wrapper.py (hunk 6 above)
Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: IMO data processing and batches should be created in a dataloader and/or a batcher that is outside of the scope of a model wrapper. Dataloaders/batchers are used in your training loop and pad your batches if they need to be padded. The model wrapper should received processed inputs and targets, otherwise the wrapper will fail to process different kind of datasets.

File: dl_training/model/model_wrapper.py (hunk 8 above)
Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: See my comment above.

Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: Not OK to close until we discuss this later.

Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: we discuss it in depth. now we can close! :) have a good day!

File: dl_training/model/model_wrapper.py (hunk 6 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: ok - after our discussion, we agreed that we can close for now - and after we merge the data loader, this part will change.
(i.e., data loader will take care of batching/padding)

File: dl_training/model/model_wrapper.py (hunk 4 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: actually, we decide to batch/pad in the data loader.
this means this code will change.

do you agree @carriepl that we can fix/change this later? after we merge the data loader?

File: dl_training/model/model_wrapper.py (hunk 5 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: same as before

Comment (carriepl) [reply dl_training/model/model_wrapper.py]: I agree about the statement that doing the batching in the loaded will fix the previous comment. However, how will doing batching in the dataloader fix this part? There is still a chance that the data loaded will give you unequal batches, no?

File: dl_training/model/model_wrapper.py (hunk 4 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: see next comment - closing

File: dl_training/model/model_wrapper.py (hunk 7 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: model.eval is enough for the correctness of the process - but no_grad will improve resource usage (both time and memory) - see https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615

so, adding the no_torch as well

Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: fixed!

File: dl_training/model/model_wrapper.py (hunk 5 above)
Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: yes, you are right indeed.. this will happen in any case. need to take care of this.
(will close the previous comment)

Comment (mirkobronzi) [reply dl_training/model/model_wrapper.py]: @carriepl  / @FrancisGregoire :

doesn't show up here - but I modifies this part of the code.
See line
https://github.com/mirkobronzi/dl_training/pull/24/files#diff-315e8aa7ba5f5cbc71790096e16607caR92

You can see now I am **not** computing the mean over the batch dimension, so that I have a loss per example. (not one loss per batch)
(like, `loss` at line 91 is a list of losses, one loss per example)
I just collect all of them (line 92), and compute the mean at the end (line 93) - by doing so, I can handle batch with different size.

Let me know if this may work for you.

Comment (FrancisGregoire) [reply dl_training/model/model_wrapper.py]: It works for me Mirko :) !

Comment (carriepl) [reply dl_training/model/model_wrapper.py]: Works for me too.

**Comment by carriepl** [comment]
Test coverage decreases but it's a negligible amount.
